src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -949,137 +952,149 @@ private unsafe bool TryGetOrCreateObjectForComInstanceInternal(
 
             using ComHolder releaseIdentity = new ComHolder(identity);
 
-            if (!flags.HasFlag(CreateObjectFlags.UniqueInstance))
+            // If the user has requested a unique instance,
+            // we will immediately create the object, register it,
+            // and return.
+            if (flags.HasFlag(CreateObjectFlags.UniqueInstance))
             {
-                using (_lock.EnterScope())
-                {
-                    if (_rcwCache.TryGetValue(identity, out GCHandle handle))
-                    {
-                        object? cachedWrapper = handle.Target;
-                        if (cachedWrapper is not null)
-                        {
-                            retValue = cachedWrapper;
-                            return true;
-                        }
-                        else
-                        {
-                            // The GCHandle has been clear out but the NativeObjectWrapper
-                            // finalizer has not yet run to remove the entry from _rcwCache
-                            _rcwCache.Remove(identity);
-                        }
-                    }
+                retValue = CreateAndRegisterObjectForComInstance(identity, inner, flags);
+                return retValue is not null;
+            }
 
-                    if (wrapperMaybe is not null)
-                    {
-                        retValue = wrapperMaybe;
-                        NativeObjectWrapper wrapper = NativeObjectWrapper.Create(
-                            identity,
-                            inner,
-                            this,
-                            retValue,
-                            flags);
-                        if (!s_rcwTable.TryAdd(retValue, wrapper))
-                        {
-                            wrapper.Release();
-                            throw new NotSupportedException();
-                        }
-                        _rcwCache.Add(identity, wrapper._proxyHandle);
-                        AddWrapperToReferenceTrackerHandleCache(wrapper);
-                        return true;
-                    }
-                }
-                if (flags.HasFlag(CreateObjectFlags.Unwrap))
+            // If we have a live cached wrapper currently,
+            // return that.
+            if (_rcwCache.FindProxyForComInstance(identity) is object liveCachedWrapper)
+            {
+                retValue = liveCachedWrapper;
+                return true;
+            }
+
+            // If the user tried to provide a pre-created managed wrapper, try to register
+            // that object as the wrapper.
+            if (wrapperMaybe is not null)
+            {
+                retValue = RegisterObjectForComInstance(identity, inner, wrapperMaybe, flags);
+                return true;
```

Comment:
`return retValue is not null;` was used above, but an explicit `true` is returned here. Is there a difference in these paths?

Reply:
No there shouldn't be. Good catch. I'll change this one.